### PR TITLE
XWIKI-23189: On smaller or half-screen browser windows, annotation boxes overflow beyond the right edge of the page

### PR DESCRIPTION
**Pull request by Sereza7**
# Jira URL

https://jira.xwiki.org/browse/XWIKI-23189

# Changes

## Description

* Added a check in the script to position the bubble on the left if there's not enough space on the right.
* Changed the position of the bubble pointer (style only element) relative to the bubble to fit the new possible position of the bubble relative to the text.
* Reworded a few comments

# Screenshots & Video

Here are a couple screenshots of comment creation bubbles with the changes proposed in this PR applied to the test instance.
Note that comment inline viewing bubbles follow the same positioning script and styles, I didn't take screenshots of them but it worked as expected.
Screenshot from 2025-07-16 16-04-57
Screenshot from 2025-07-16 16-04-38

# Executed Tests

Manual tests, see above. In manual tests, it appeared that using 560px for the treshhold resulted in some unexpected edge cases where the box could overflow the page slightly. That's why I decided to make the checks a bit wider, even if it meant losing a few cases where the box would fit on the right. IMO having the box line up with the very side of the screen looked off anyways.

Successfully built, in this order:
* `mvn clean install -f xwiki-platform-core/xwiki-platform-annotation/xwiki-platform-annotation-ui/ -Pquality`
* `mvn clean install -f xwiki-platform-core/xwiki-platform-annotation/xwiki-platform-annotation-test/xwiki-platform-annotation-test-pageobjects`
* `mvn clean install -f xwiki-platform-core/xwiki-platform-annotation/xwiki-platform-annotation-test/xwiki-platform-annotation-test-docker/`

# Expected merging strategy

* Prefers squash: Yes 
* Backport on branches:
  * 17.4.X
  * NOT 16.10.X because even though it's a bugfix, it's a very old one with quite a low impact on usability, with DOM changes that might be dangerous.

### CODE REVIEW

File: xwiki-platform-core/xwiki-platform-annotation/xwiki-platform-annotation-ui/src/main/resources/AnnotationCode/Style.xml [hunk 1]
```diff
@@ -289,12 +289,12 @@
 .annotation-bubble {
   background: $theme.backgroundSecondaryColor;
   border: 1px dotted $theme.borderColor;
-  width: 560px;
+  width: 560px; /* Take care when updating this hard-coded value, script on AnnotationCode.Script relies on it. */
```
Comment (michitux) [inline xwiki-platform-core/xwiki-platform-annotation/xwiki-platform-annotation-ui/src/main/resources/AnnotationCode/Style.xml]: I think we normally put comments on their own line. Also, it would be nice if the JavaScript code could get the real width of the element instead of relying on hard-coded values. Switching positions based on a fixed value is of course a first step, using computed values could be left as a future improvement.

File: xwiki-platform-core/xwiki-platform-annotation/xwiki-platform-annotation-ui/src/main/resources/AnnotationCode/Script.xml [hunk 2]
```diff
@@ -1362,18 +1362,27 @@ XWiki.Annotation = Class.create({
...
-    bubble.style.top = top + 'px';
+    // position the bubble
+    // By default, we position it to the right and under the coordinates
+    // Position it horizontally. The width of the bubble is fixed at 560px to which we add a 20px left margin
+    if ((horizontalCoordinate &lt; 580) || (horizontalCoordinate + 580 &lt; window.innerWidth) ) {
+      bubble.style.left = horizontalCoordinate + 'px';
+    }
+    else {
+      // There isn't enough space on the right of the window to fit the modal, and there is enough on the left.
+      bubble.style.right = (window.innerWidth-horizontalCoordinate) + 'px';
```
Comment (michitux) [inline xwiki-platform-core/xwiki-platform-annotation/xwiki-platform-annotation-ui/src/main/resources/AnnotationCode/Script.xml]:
```suggestion
      bubble.style.right = (window.innerWidth - horizontalCoordinate) + 'px';
```

As bubbles aren't re-positioned when the window is resized, I'm wondering if positioning from the left wouldn't be more stable/less surprising as at least all bubbles would behave the same and not move in different ways.

File: xwiki-platform-core/xwiki-platform-annotation/xwiki-platform-annotation-ui/src/main/resources/AnnotationCode/Script.xml [hunk 3]
```diff
@@ -1362,18 +1362,27 @@ XWiki.Annotation = Class.create({
...
-    // make it hidden for the moment
+    // keep the bubble hidden for the moment
     bubble.toggleClassName('hidden');
-    // position it
-    bubble.style.left = left + 'px';
-    bubble.style.top = top + 'px';
+    // position the bubble
+    // By default, we position it to the right and under the coordinates
+    // Position it horizontally. The width of the bubble is fixed at 560px to which we add a 20px left margin
+    if ((horizontalCoordinate &lt; 580) || (horizontalCoordinate + 580 &lt; window.innerWidth) ) {
```
Comment (michitux) [inline xwiki-platform-core/xwiki-platform-annotation/xwiki-platform-annotation-ui/src/main/resources/AnnotationCode/Script.xml]:
```suggestion
    if (horizontalCoordinate &lt; 580 || horizontalCoordinate + 580 &lt; window.innerWidth) {
```

File: xwiki-platform-core/xwiki-platform-annotation/xwiki-platform-annotation-ui/src/main/resources/AnnotationCode/Script.xml [hunk 4]
```diff
@@ -1362,18 +1362,27 @@ XWiki.Annotation = Class.create({
...
+      bubble.style.left = horizontalCoordinate + 'px';
+    }
+    else {
```
Comment (michitux) [inline xwiki-platform-core/xwiki-platform-annotation/xwiki-platform-annotation-ui/src/main/resources/AnnotationCode/Script.xml]:
```suggestion
    } else {
```

File: xwiki-platform-core/xwiki-platform-annotation/xwiki-platform-annotation-ui/src/main/resources/AnnotationCode/Style.xml (hunk 1 above)
Comment (Sereza5) [reply xwiki-platform-core/xwiki-platform-annotation/xwiki-platform-annotation-ui/src/main/resources/AnnotationCode/Style.xml]: I used a computed value as it was pretty straightforward to do. The only "compromise" I had to do was that I had to change the order of operations. Now, the bubble is made visible before it's positioned. I checked how it looked and I couldn't see any flicker when loading the bubble, so it's not much of a compromise :)

____
Removed the comment because it does not apply anymore since https://github.com/xwiki/xwiki-platform/pull/4382/commits/f75eea473fd7c99c79c0381effe4c58c2c91315d :+1:

File: xwiki-platform-core/xwiki-platform-annotation/xwiki-platform-annotation-ui/src/main/resources/AnnotationCode/Script.xml (hunk 2 above)
Comment (Sereza5) [reply xwiki-platform-core/xwiki-platform-annotation/xwiki-platform-annotation-ui/src/main/resources/AnnotationCode/Script.xml]: Note that at least for the bubbles opened on the very left side of the screen, we'd still need to place them on the right.

File: xwiki-platform-core/xwiki-platform-annotation/xwiki-platform-annotation-ui/src/main/resources/AnnotationCode/Script.xml (hunk 4 above)
Comment (Sereza5) [reply xwiki-platform-core/xwiki-platform-annotation/xwiki-platform-annotation-ui/src/main/resources/AnnotationCode/Script.xml]: Addressed in https://github.com/xwiki/xwiki-platform/pull/4382/commits/f75eea473fd7c99c79c0381effe4c58c2c91315d :+1:

File: xwiki-platform-core/xwiki-platform-annotation/xwiki-platform-annotation-ui/src/main/resources/AnnotationCode/Script.xml (hunk 3 above)
Comment (Sereza5) [reply xwiki-platform-core/xwiki-platform-annotation/xwiki-platform-annotation-ui/src/main/resources/AnnotationCode/Script.xml]: Addressed in https://github.com/xwiki/xwiki-platform/commit/f75eea473fd7c99c79c0381effe4c58c2c91315d and https://github.com/xwiki/xwiki-platform/pull/4382/commits/08f36b4d4db9d52d4fb139fb8c8a380fd4cf6855 👍

File: xwiki-platform-core/xwiki-platform-annotation/xwiki-platform-annotation-ui/src/main/resources/AnnotationCode/Script.xml (hunk 2 above)
Comment (michitux) [reply xwiki-platform-core/xwiki-platform-annotation/xwiki-platform-annotation-ui/src/main/resources/AnnotationCode/Script.xml]: My point was that when the window is resized, currently all bubbles stay at the same position relative to the top left corner of the window. With this change, the bubbles that are positioned from the right edge would stay at the same position relative to the top right corner of the screen, creating an inconsistency. The best would of course be if all of them stayed at the position where they actually belonged. I don't see any problem with bubbles at the very left of the screen, what would be the problem with them?

File: xwiki-platform-core/xwiki-platform-annotation/xwiki-platform-annotation-ui/src/main/resources/AnnotationCode/Script.xml [hunk 5]
```diff
@@ -1362,22 +1362,29 @@ XWiki.Annotation = Class.create({
...
-    // make it visible
-    bubble.toggleClassName('hidden');
     // put this bubble in the bubbles stack
     this.bubbles.push(bubble);
+    // position the bubble
+    // By default, we position it to the right and under the coordinates
+    // Position it horizontally.
+    let bubbleWidth = bubble.offsetWidth 
+      + parseInt(getComputedStyle(bubble).marginLeft) 
+      + parseInt(getComputedStyle(bubble).marginRight);
```
Comment (michitux) [inline xwiki-platform-core/xwiki-platform-annotation/xwiki-platform-annotation-ui/src/main/resources/AnnotationCode/Script.xml]: Using `getComputedStyle()` twice seems bad in terms of performance (and easy to avoid). Also, wouldn't it be simpler to use [`getBoundingClientRect`](https://developer.mozilla.org/en-US/docs/Web/API/Element/getBoundingClientRect)? I'm also wondering if it wouldn't be safer to avoid flickering to set the `left` style to something like `-9999px` to position the element outside the screen before showing it.

Comment (Sereza5) [reply xwiki-platform-core/xwiki-platform-annotation/xwiki-platform-annotation-ui/src/main/resources/AnnotationCode/Script.xml]: > I'm also wondering if it wouldn't be safer to avoid flickering to set the left style to something like -9999px to position the element outside the screen before showing it.

I couldn't see any flickering running it on my machine. The bubble is inserted in the doc at line 1371 and the position changed at line 1381.
Screenshot from 2025-08-26 13-59-11
Screenshot from 2025-08-26 13-59-22

Here are the two positions, and most of the time the users won't even see under the footer bar. IMO this is an alright flicker to keep.
Note that we cannot get the BoundingClientRect or any style data from the bubble if it's not yet in the DOM. This means that we cannot just set its position before creating it.

> Also, wouldn't it be simpler to use [getBoundingClientRect](https://developer.mozilla.org/en-US/docs/Web/API/Element/getBoundingClientRect)?

`getBoundingClientRect` calculates the width without margins. Since we use the margins to position the "arrow" on the side of the bubble, it's important we count those in. In order to retrieve margins, the most stable solution I could find was to use the computedStyle. We could technically hardcode the margin value since I doubt it changes much, but retrieving the style makes sure  style customizations still work as expected here.

Without factoring in these margins there's a 20px column where if the user adds the annotation, the bubble will go over the right of the viewport.

With the margins, the bubble goes up to here:
image

Then is moved to the left of the click position.

Comment (Sereza5) [reply xwiki-platform-core/xwiki-platform-annotation/xwiki-platform-annotation-ui/src/main/resources/AnnotationCode/Script.xml]: I added the factorization of the calls in https://github.com/xwiki/xwiki-platform/pull/4382/commits/86aeddb895d71d521ca24e909268608e1c00d7b9 :+1:

File: xwiki-platform-core/xwiki-platform-annotation/xwiki-platform-annotation-ui/src/main/resources/AnnotationCode/Script.xml (hunk 2 above)
Comment (Sereza5) [reply xwiki-platform-core/xwiki-platform-annotation/xwiki-platform-annotation-ui/src/main/resources/AnnotationCode/Script.xml]: I didn't understand your point in the first message, now I understand, thank you for sharing details.

Fixed in https://github.com/xwiki/xwiki-platform/pull/4382/commits/86aeddb895d71d521ca24e909268608e1c00d7b9 :+1:

**Comment by Sereza7** [comment]
This PR is ready for further reviewing and/or a fix :)

File: xwiki-platform-core/xwiki-platform-annotation/xwiki-platform-annotation-ui/src/main/resources/AnnotationCode/Style.xml [hunk 6]
```diff
@@ -316,6 +316,28 @@
   border-right-color: $theme.backgroundSecondaryColor;
 }
 
+.annotation-bubble.annotation-bubble-position-left {
+  border-top-left-radius: 10px;
+  border-top-right-radius: 0;
+  
+  &amp;:before, &amp;:after {
```
Comment (michitux) [inline xwiki-platform-core/xwiki-platform-annotation/xwiki-platform-annotation-ui/src/main/resources/AnnotationCode/Style.xml:323]: I think it's okay, but I just wanted to mention that this is the first time that I notice that we use CSS nesting (this is CSS, not LESS from what I can see) to ensure that this has been intentional.

**Review by michitux — APPROVED** [review]
Thank you very much! I'm not fully convinced regarding the flickering, as this could add and remove again a scrollbar and might (not sure) change the scrolling position, both not very nice. The solution I see is to:

1. Set the correct vertical position before displaying as we know that already.
2. Set the `left` property to `left: -999em` before displaying to move the element off-screen without introducing any scrollbars.

This should be a small change and should completely avoid the risk of any flickering.

File: xwiki-platform-core/xwiki-platform-annotation/xwiki-platform-annotation-ui/src/main/resources/AnnotationCode/Style.xml (hunk 6 above)
Comment (Sereza5) [reply xwiki-platform-core/xwiki-platform-annotation/xwiki-platform-annotation-ui/src/main/resources/AnnotationCode/Style.xml:323]: I'm surprised this is only the first time.

https://caniuse.com/css-nesting

CSS Nesting is baseline 2023 and this kind of use of it is supported for about 90% of users. 
I believe this was one of the reasons why we decided it was okay to start moving from LESS to native CSS so I don't think it needs further validation to start using it.
